### DISCUSSION TITLE
run inference on GPU or MPS

### DISCUSSION THREAD

**Discussion by mbui0327** [pyannote/pyannote-audio#1155] [General]
I could not find any document on running the diarization on GPU or MPS. The earlier version (v1) can be loaded into a torch pipeline and can be assigned device to it. Is there equivalent thing in v2.
Thanks

**Comment by hbredin** [comment]
Setting the device used in pipelines manually is not currently supported.
pyannote.audio will automatically use GPU (not MPS) when available.
That being said, I'd be happy to look at a pull request adding this feature, if someone (you?) comes up with one.

**Reply by mbui0327** [reply]
Can you tell me where I should look into to do this? Took a quick look at `audio/core/pipeline.py` but could not find out how it is connected to torch pipeline. Thanks.

**Reply by hbredin** [reply]
https://github.com/pyannote/pyannote-audio/blob/d966b518b0d6383c827a5e9a978516652f871a27/pyannote/audio/pipelines/speaker_diarization.py#L141

**Comment by fablau** [comment]
If we can define the device, we could use the new pytorch optimized for M1 on Macs as explained here:

https://towardsdatascience.com/installing-pytorch-on-apple-m1-chip-with-gpu-acceleration-3351dc44d67c

**Comment by mbui0327** [comment]
Yes, it is pretty trivial. I have it working. Will make a PR later this
week.

~Bach

**Comment by fablau** [comment]
Awesome! Thank you!

**Comment by fablau** [comment]
@mbui0327 Any news about all this?

**Comment by hbredin** [comment]
[This PR](https://github.com/pyannote/pyannote-audio/pull/1204) adds a new `pipeline.to(device)` API. Would you mind testing it?

**Comment by fablau** [comment]
Thank you! It works great! It is much, much faster! 20 seconds vs 3 minutes for a 40 seconds dialog. What an improvement! I just need to analyze the results because for some reason they seem to be different... I'll be back with more info.

**Comment by fablau** [comment]
Yeah, for some reason, I get screwed results using the mps device. Is the code below the correct way to use it?

```
from pyannote.audio import Pipeline

pipeline = Pipeline.from_pretrained("pyannote/speaker-diarization", use_auth_token="[my_token]")

pipeline.to("mps")

diarization = pipeline("audio_dialog.wav")

for turn, _, speaker in diarization.itertracks(yield_label=True):
	print(f"start={turn.start:.1f}s stop={turn.end:.1f}s speaker_{speaker}")
```

With the CPU, which takes over 3 minutes to process, I get the following correct results:

```
start=1.4s stop=13.3s speaker_SPEAKER_00
start=13.3s stop=15.3s speaker_SPEAKER_01
start=15.8s stop=32.0s speaker_SPEAKER_00
start=32.4s stop=37.5s speaker_SPEAKER_01
```

With the GPU (using pipeline.to("mps")) which takes just 20 seconds, I get the following results:

```
start=36.9s stop=37.5s speaker_SPEAKER_00
```

Any ideas?

**Comment by fablau** [comment]
Well, after some research, I think the problem was that I didn't have PyTorch Nightly installed on my Mac.  Fact is, after having created a new conda environment from scratch with the Nightly Pytorch on it, by performing:

```
pip3 install pyannote.audio
```

It seems to downgrade pytorch audio and other stuff so that once I execute the python script above, I get a bunch of errors:

```
/opt/homebrew/Caskroom/miniconda/base/envs/pyannote6/lib/python3.8/site-packages/torchaudio/_internal/module_utils.py:99: UserWarning: Failed to import soundfile. 'soundfile' backend is not available.
  warnings.warn("Failed to import soundfile. 'soundfile' backend is not available.")
Traceback (most recent call last):
  File "diary.py", line 8, in <module>
    from pyannote.audio import Pipeline
  File "/opt/homebrew/Caskroom/miniconda/base/envs/pyannote6/lib/python3.8/site-packages/pyannote/audio/__init__.py", line 29, in <module>
    from .core.inference import Inference
  File "/opt/homebrew/Caskroom/miniconda/base/envs/pyannote6/lib/python3.8/site-packages/pyannote/audio/core/inference.py", line 34, in <module>
    from pyannote.audio.core.io import AudioFile
  File "/opt/homebrew/Caskroom/miniconda/base/envs/pyannote6/lib/python3.8/site-packages/pyannote/audio/core/io.py", line 38, in <module>
    import torchaudio
  File "/opt/homebrew/Caskroom/miniconda/base/envs/pyannote6/lib/python3.8/site-packages/torchaudio/__init__.py", line 1, in <module>
    from torchaudio import (  # noqa: F401
  File "/opt/homebrew/Caskroom/miniconda/base/envs/pyannote6/lib/python3.8/site-packages/torchaudio/_extension.py", line 135, in <module>
    _init_extension()
  File "/opt/homebrew/Caskroom/miniconda/base/envs/pyannote6/lib/python3.8/site-packages/torchaudio/_extension.py", line 105, in _init_extension
    _load_lib("libtorchaudio")
  File "/opt/homebrew/Caskroom/miniconda/base/envs/pyannote6/lib/python3.8/site-packages/torchaudio/_extension.py", line 52, in _load_lib
    torch.ops.load_library(path)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/pyannote6/lib/python3.8/site-packages/torch/_ops.py", line 640, in load_library
    ctypes.CDLL(path)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/pyannote6/lib/python3.8/ctypes/__init__.py", line 373, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: dlopen(/opt/homebrew/Caskroom/miniconda/base/envs/pyannote6/lib/python3.8/site-packages/torchaudio/lib/libtorchaudio.so, 0x0006): Symbol not found: (__ZN2at4_ops9_pad_enum4callERKNS_6TensorEN3c108ArrayRefIxEExNS5_8optionalIdEE)
  Referenced from: '/opt/homebrew/Caskroom/miniconda/base/envs/pyannote6/lib/python3.8/site-packages/torchaudio/lib/libtorchaudio.so'
  Expected in: '/opt/homebrew/Caskroom/miniconda/base/envs/pyannote6/lib/python3.8/site-packages/torch/lib/libtorch_cpu.dylib'
```

Is pyannote compatible with PyTorch Nightly?

**Reply by adeviney** [reply]
@fablau I am having the same issue, and I am wondering -- why do you think the problem is (the lack of) PyTorch Nightly? 🤔 

On my M1 mac, I am getting the same results you are after installing pyannote.audio as the current head of the develop branch and using `pipeline.to("mps")` (the pipeline being fit _much_ faster, but the entire audio file being attributed to speaker 0).
However, the same thing also happened with Google Colab and their CUDA GPU.

So I am wondering if there is another issue?

**Reply by fablau** [reply]
This is interesting @adeviney ... are you using the `pipeline.to("mps")` there too? If so, that's probably why... if you use a GPU, don't set the "mps" setting. And the fact you get the same kind of problem makes me think that maybe the problem is not an incompatibility with PyTorch Nightly but with the mps pipeline?

Could @hbredin chime in and comment on this?

**Reply by adeviney** [reply]
Good point, although I'm pretty sure I used `pipeline.to("cuda")` for the Google colab test (or commented out that line altogether, assuming it still defaults to gpu if it's available). But I will confirm and see if I can figure it out later. :-)

**Comment by petros94** [comment]
Has anyone managed to run pyannote on mps? 🤔

**Reply by fablau** [reply]
I haven't tried lately...

**Comment by IKabitz** [comment]
Also wondering about this. Running into issues with diarization results when using torch.device('mps'). @fablau @adeviney Did either of you find a solution to your issues?

**Reply by fablau** [reply]
I haven't tried that in a while. I'll try again as soon as possible.

**Reply by dasqueel** [reply]
Same here.  It would be nice to use my new MacBook Pro to do some ML stuff lol.  I remember it was advertised as proficient in ML computing.

Update:
```
NotImplementedError: The operator 'aten::_fft_r2c' is not currently implemented for the MPS device. If you want this op to be added in priority during the prototype phase of this feature, please comment on https://github.com/pytorch/pytorch/issues/77764. As a temporary fix, you can set the environment variable `PYTORCH_ENABLE_MPS_FALLBACK=1` to use the CPU as a fallback for this op. WARNING: this will be slower than running natively on MPS.
```

Looks like an implementation needs to be made?

**Reply by nahuel89p** [reply]
I think it's this? https://github.com/pytorch/pytorch/issues/78044

**Comment by nahuel89p** [comment]
Isn't it this issue?
https://github.com/pytorch/pytorch/issues/78044

**Comment by jbeltran73-2** [comment]
I have found this post that maybe would help

https://medium.com/buzzwordplus/speech-to-text-unlocking-speech-recognition-with-apples-machine-learning-framework-mlx-08416acbeb77

**Comment by agileandy** [comment]
I think I've got this working...

import torch and then set the pipeline to use the torch device mps. 

```
import torch
from pyannote.audio import Pipeline
pipeline = Pipeline.from_pretrained("pyannote/speaker-diarization@2.1",
                                    use_auth_token="hf........")

pipeline.to(torch.device("mps"))

diarization = pipeline("audio.wav")


# dump the diarization output to disk using RTTM format
with open("audio.rttm", "w") as rttm:
    diarization.write_rttm(rttm)
```

**Reply by gammieduncan** [reply]
That "works" but it doesn't actually run all the pytorch operations pyannote uses on your MPS, meaning that there will still be a significant slowdown compared to what you would have on, say, an Nvidia GPU. This appears to be a problem on the pytorch side of things, where a decent chunk of the operators aren't yet implemented for mps.

**Comment by andrew867** [comment]
Honestly unsure, but this may help? https://github.com/andrew867/pyannote-openvino
